reactos/drivers/usb/usbport/roothub.c

```diff
@@ -0,0 +1,1019 @@
+
+        EndpointDescriptor->bLength = sizeof(USB_ENDPOINT_DESCRIPTOR);
+        EndpointDescriptor->bDescriptorType = USB_ENDPOINT_DESCRIPTOR_TYPE;
+        EndpointDescriptor->bEndpointAddress = 0x00;
+        EndpointDescriptor->bmAttributes = 0x00;
+        EndpointDescriptor->wMaxPacketSize = 0x0040;
+        EndpointDescriptor->bInterval = 0x00;
+
+        Status = USBPORT_OpenPipe(FdoDevice,
+                                  DeviceHandle,
+                                  &DeviceHandle->PipeHandle,
+                                  NULL);
+    }
+    else
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    return Status;
+}
+
+ULONG
+NTAPI
+USBPORT_InvalidateRootHub(PVOID Context)
+{
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+    PDEVICE_OBJECT FdoDevice;
+    PDEVICE_OBJECT PdoDevice;
+    PUSBPORT_RHDEVICE_EXTENSION PdoExtension;
+    PUSBPORT_ENDPOINT Endpoint = NULL;
+
+    DPRINT("USBPORT_InvalidateRootHub ... \n");
+
+    FdoExtension = (PUSBPORT_DEVICE_EXTENSION)((ULONG_PTR)Context -
+                                               sizeof(USBPORT_DEVICE_EXTENSION));
+
+    FdoDevice = FdoExtension->CommonExtension.SelfDevice;
+
+    if (FdoExtension->Flags & USBPORT_FLAG_HC_SUSPEND && 
+         FdoExtension->Flags & USBPORT_FLAG_HC_WAKE_SUPPORT &&
+         FdoExtension->MiniPortFlags & USBPORT_MPFLAG_SUSPENDED &&
+         FdoExtension->TimerFlags & USBPORT_TMFLAG_WAKE)
+    {
+        USBPORT_HcQueueWakeDpc(FdoDevice);
+        return 0;
+    }
+    
+    FdoExtension->MiniPortInterface->Packet.RH_DisableIrq(FdoExtension->MiniPortExt);
+
+    PdoDevice = FdoExtension->RootHubPdo;
+
+    if (PdoDevice)
+    {
+        PdoExtension = PdoDevice->DeviceExtension;
+        Endpoint = PdoExtension->Endpoint;
+    }
+
+    if (Endpoint)
+    {
+        USBPORT_InvalidateEndpointHandler(FdoDevice,
+                                          PdoExtension->Endpoint,
+                                          INVALIDATE_ENDPOINT_WORKER_THREAD);
+    }
+
+    return 0;
+}
+
+VOID
+NTAPI
+USBPORT_RootHubPowerAndChirpAllCcPorts(IN PDEVICE_OBJECT FdoDevice)
+{
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+    PUSBPORT_REGISTRATION_PACKET Packet;
+    USBPORT_ROOT_HUB_DATA RootHubData;
+    ULONG Port;
+    PDEVICE_RELATIONS CompanionControllersList;
+    PUSBPORT_DEVICE_EXTENSION CompanionFdoExtension;
+    PUSBPORT_REGISTRATION_PACKET CompanionPacket;
+    ULONG CompanionPorts;
+    ULONG NumController;
+    PDEVICE_OBJECT * Entry;
+    ULONG NumPorts;
+
+    DPRINT("USBPORT_RootHub_PowerAndChirpAllCcPorts: FdoDevice - %p\n",
+           FdoDevice);
+
+    FdoExtension = FdoDevice->DeviceExtension;
+
+    Packet = &FdoExtension->MiniPortInterface->Packet;
+
+    RtlZeroMemory(&RootHubData, sizeof(RootHubData));
+
+    Packet->RH_GetRootHubData(FdoExtension->MiniPortExt,
+                              &RootHubData);
+
+    NumPorts = RootHubData.NumberOfPorts;
+
+    for (Port = 1; Port < NumPorts; ++Port)
```

Comment:
Double-check this? I would expect `for (Port = 1; Port <= NumPorts; ++Port)` for 1-based port numbers and `for (port = 0; Port < NumPorts; ++Port)` for zero-based